201911012 Yeong jae Kong

(1) For MNIST data set, train Logistic regression models and find the best model that can achieve
the highest accuracy on the test data set. Provide the average & standard deviation of all test data.

In [22]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

dataset = datasets.load_digits()
x_data = dataset.data
y_data = dataset.target

x_train, x_test, y_train, y_test = model_selection.train_test_split(x_data, y_data, test_size=0.3)

#general C
model = LogisticRegression(max_iter=5000, tol=0.0001, penalty='l2', C=1.0)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
print(accuracy)

#low C
model_2 = LogisticRegression(max_iter=5000, tol=0.0001, penalty='l2', C=0.01)
model_2.fit(x_train, y_train)
y_predict_2 = model_2.predict(x_test) 
accuracy_2 = metrics.accuracy_score(y_test, y_predict_2)
print(accuracy_2)

#high C
model_3 = LogisticRegression(max_iter=5000, tol=0.0001, penalty='l2', C=100)
model_3.fit(x_train, y_train)
y_predict_3 = model_3.predict(x_test) 
accuracy_3 = metrics.accuracy_score(y_test, y_predict_3)
print(accuracy_3)

0.9537037037037037
0.9629629629629629
0.9537037037037037


Thus, if C parameter is low, accuracy increases.
When C=0.01 shows the highest accuracy(0.9630).

(2) For the same data set, train K-NN classifiers and find the best model that can achieve the highest
accuracy on the test data set. Provide the average & standard deviation of all test data.

In [36]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier(n_neighbors=5, n_jobs=-1)
model.fit(x_train, y_train)
y_predict = model.predict(x_test)
accuracy = metrics.accuracy_score(y_test, y_predict)

for k in range(1, 11):
    model = KNeighborsClassifier(n_neighbors=k, n_jobs=-1)
    model.fit(x_train, y_train)
    accuracy = model.score(x_test, y_test)
    print('k: {}, accuracy: {}'.format(k, accuracy))

k: 1, accuracy: 0.9888888888888889
k: 2, accuracy: 0.9796296296296296
k: 3, accuracy: 0.987037037037037
k: 4, accuracy: 0.9851851851851852
k: 5, accuracy: 0.9833333333333333
k: 6, accuracy: 0.9814814814814815
k: 7, accuracy: 0.9833333333333333
k: 8, accuracy: 0.9814814814814815
k: 9, accuracy: 0.9814814814814815
k: 10, accuracy: 0.9796296296296296


k=1 model can achieve the highest accuracy(0.9888) on the test data set.

(3) For the same data set, train SVM classifiers and find the best model that can achieve the highest
accuracy on the test data set. Provide the average & standard deviation of all test data.

In [67]:
from sklearn.svm import SVC
model = SVC(kernel='linear', C=1.0)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
#accuracy = 0.9703703703703703
model = SVC(kernel='rbf', C=1.0)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
#accuracy = 0.9796296296296296

#rbf's accuracy is higher than linear, so we use rbf
#then modulate C

model = SVC(kernel='rbf', C=1.0)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
print(accuracy)
model = SVC(kernel='rbf', C=10)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
print(accuracy)
model = SVC(kernel='rbf', C=0.1)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
print(accuracy)

0.9796296296296296
0.9833333333333333
0.9444444444444444


When I use linear SVC, accuracy is 0.9704 but rbf SVC's accuracy is 0.9796.
SVM model that kernel='rbf' and C = 10 can achieve the highest accuracy on the test data set.(0.9833)

(4) For the same data set, train Random forest classifiers and find the best model that can achieve the highest accuracy on the test data set. Provide the average & standard deviation of all test data.

In [79]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

#Search parameter that show the highest accuracy by GridSearchCV

params = { 'n_estimators' : [10, 100],
           'max_depth' : [4, 6, 8, 10, 12],
           'min_samples_leaf' : [3, 5, 7],
           'min_samples_split' : [2, 3, 5]
            }

#The higher the n_estimators is the better, but it takes a long time, so adjust it appropriately.

rf_clf = RandomForestClassifier(random_state = 0, n_jobs = -1)
grid_cv = GridSearchCV(rf_clf, param_grid = params, cv = 3, n_jobs = -1)
grid_cv.fit(x_train, y_train)

print('Optimal hyper parameter: ', grid_cv.best_params_)
print('Highest prediction accuracy: {:.6f}'.format(grid_cv.best_score_))

model = RandomForestClassifier(max_depth=8, min_samples_leaf=3, min_samples_split=2, n_estimators=100)
model.fit(x_train, y_train)
y_predict = model.predict(x_test) 
accuracy = metrics.accuracy_score(y_test, y_predict)
print(accuracy)

Optimal hyper parameter:  {'max_depth': 8, 'min_samples_leaf': 3, 'min_samples_split': 2, 'n_estimators': 100}
Highest prediction accuracy: 0.964200
0.9703703703703703


As a result of comparing parameter values through GridsearchCV, the optimal parameter values that show highest accuracy are 'max_depth': 8,'min_samples_leaf': 3,'min_samples_split': 2, and'n_estimators': 100. the accuracy is 0.9704

| - |Logistic regression|KNN classifiers|SVM|RandomForestClassifier|
|---|---|---|---|---|
|Accuracy|96.30%|98.88%|98.33%|97.04%|